trio/tests/test_file_io.py

```diff
@@ -0,0 +1,169 @@
+import io
+import tempfile
+
+import pytest
+from unittest import mock
+from unittest.mock import patch, sentinel
+
+import trio
+from trio import _core
+from trio._file_io._file_io import _FILE_SYNC_ATTRS, _FILE_ASYNC_METHODS
+
+
+@pytest.fixture
+def path(tmpdir):
+    return tmpdir.join('test').__fspath__()
+
+
+@pytest.fixture
+def wrapped():
+    return mock.Mock(spec_set=io.StringIO)
+
+
+@pytest.fixture
+def async_file(wrapped):
+    return trio.wrap_file(wrapped)
+
+
+def test_wrap_invalid():
+    with pytest.raises(TypeError):
+        trio.wrap_file(str())
+
+
+def test_wrapped_property(async_file, wrapped):
+    assert async_file.wrapped == wrapped
+
+
+def test_dir_matches_wrapped(async_file, wrapped):
+    attrs = _FILE_SYNC_ATTRS + _FILE_ASYNC_METHODS
+
+    # all supported attrs in wrapped should be available in async_file
+    assert all(attr in dir(async_file) for attr in attrs if attr in dir(wrapped))
+    # all supported attrs not in wrapped should not be available in async_file
+    assert not any(attr in dir(async_file) for attr in attrs if attr not in dir(wrapped))
```

Comment:
> So like if StringIO actually defined all the IO attributes, then the second line would fail to test anything, right?

Well yes. If we wanted to be more rigorous I could test all three cases (having none of the supported attributes implemented, having some of them, and having all of them).